File: test-code/src/test/java/com/example/testcode/member/MemberApiTest.java
package com.example.testcode.member;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.micrometer.core.instrument.util.IOUtils;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.core.io.ResourceLoader;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@RunWith(SpringRunner.class)
@AutoConfigureMockMvc
@Transactional
public class MemberApiTest {

  @Autowired
  private MockMvc mvc;

  @Autowired
  protected ObjectMapper objectMapper;

  @Autowired
  protected ResourceLoader resourceLoader;

  @Test
  public void 회원가입테스트() throws Exception {

    //given
    final MemberSingUpRequest dto = new MemberSingUpRequest("yun", "yun@asd.com");

    //when
    final ResultActions resultActions = mvc.perform(post("/members")
        .contentType(MediaType.APPLICATION_JSON_UTF8)
        .content(objectMapper.writeValueAsString(dto)))
        .andDo(print());

    //then
    resultActions
        .andExpect(status().isOk());

  }


  @Test
  public void json_파일로테스트() throws Exception {
    //given
    final String requestBody = readJson("classpath:member-singup.json");

    //when
    final ResultActions resultActions = mvc.perform(post("/members")
        .contentType(MediaType.APPLICATION_JSON_UTF8)
        .content(requestBody))
        .andDo(print());

    //then
    resultActions
        .andExpect(status().isOk());

  }

  @Test
  public void default_접근지시자_를이용한_테스트() throws Exception {
    //given
    final MemberSingUpRequest dto = MemberSignUpRequestBuilder.build("yun", "yun@asd.com");

    //when
    final ResultActions resultActions = mvc.perform(post("/members")
        .contentType(MediaType.APPLICATION_JSON_UTF8)
        .content(objectMapper.writeValueAsString(dto)))
        .andDo(print());

    //then
    resultActions
        .andExpect(status().isOk());

  }

  @Test
  public void member_page_test() throws Exception {
    //given
    //dataSetUp("classpath:member_page_test.sql"); 로직은 없습니다.

    //when
    final ResultActions resultActions = mvc.perform(get("/members")
        .contentType(MediaType.APPLICATION_JSON_UTF8))
        .andDo(print());

    //then
    resultActions
        .andExpect(status().isOk())
        .andExpect(jsonPath("content").exists())
        .andExpect(jsonPath("pageable").exists())
        .andExpect(jsonPath("pageable").exists())
        .andExpect(jsonPath("numberOfElements").value("4"));

  }

  protected String readJson(final String path) throws IOException {
    final InputStream inputStream = resourceLoader.getResource(path).getInputStream();
    return IOUtils.toString(inputStream, StandardCharsets.UTF_8);
  }
}

File: kotlin-junit5/src/test/kotlin/com/example/kotlinjunit5/member/MemberApiTest.kt
package com.example.kotlinjunit5.member

import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.TestConstructor
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get

@SpringBootTest
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
@ActiveProfiles("test")
@AutoConfigureMockMvc
internal class MemberApiTest(
        val memberRepository: MemberRepository,
        val mockMvc: MockMvc
) {

    @Test
    internal fun `members 조회 테스트`() {
        memberRepository.saveAll(listOf(
                Member("email1@asd.com", "jin"),
                Member("email2@asd.com", "yun"),
                Member("email3@asd.com", "wan"),
                Member("email4@asd.com", "kong"),
                Member("email5@asd.com", "joo")
        ))

        mockMvc.get("/members") {
            accept = MediaType.APPLICATION_JSON
        }.andExpect {
            content { contentType(MediaType.APPLICATION_JSON) }
            jsonPath("$[0].name") { value("jin") }
            jsonPath("$[1].name") { value("yun") }
            jsonPath("$[2].name") { value("wan") }
            jsonPath("$[3].name") { value("kong") }
            jsonPath("$[4].name") { value("joo") }
        }.andDo {
            print()
        }
    }
}

//given(memberRepository.findAll()).willReturn(listOf(
//                Member("asd1@asd.com", "name1"),
//                Member("asd2@asd.com", "name2"),
//                Member("asd3@asd.com", "name3"),
//                Member("asd4@asd.com", "name4")
//        ))